bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/SecureHttpClientFactory.java

```diff
@@ -0,0 +1,219 @@
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.eclipse.smarthome.io.net.http.HttpClientFacade;
+import org.eclipse.smarthome.io.net.http.HttpClientFactory;
+import org.eclipse.smarthome.io.net.http.TrustManagerProvider;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Factory class to create jetty http clients
+ *
+ * @author Michael Bock - initial API
+ */
+@Component(service = HttpClientFactory.class, immediate = true, configurationPid = "org.eclipse.smarthome.HttpClientFactory")
+public class SecureHttpClientFactory implements HttpClientFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(SecureHttpClientFactory.class);
+
+    private static final String CONFIG_MIN_THREADS_SHARED = "minThreadsShared";
+    private static final String CONFIG_MAX_THREADS_SHARED = "maxThreadsShared";
+    private static final String CONFIG_KEEP_ALIVE_SHARED = "keepAliveTimeoutShared";
+    private static final String CONFIG_MIN_THREADS_CUSTOM = "minThreadsCustom";
+    private static final String CONFIG_MAX_THREADS_CUSTOM = "maxThreadsCustom";
+    private static final String CONFIG_KEEP_ALIVE_CUSTOM = "keepAliveTimeoutCustom";
+
+    private static final int MIN_CONSUMER_NAME_LENGTH = 4;
+    private static final int MAX_CONSUMER_NAME_LENGTH = 20;
+    private static final Pattern CONSUMER_NAME_PATTERN = Pattern.compile("[a-zA-Z0-9_\\-]*");
+
+    private volatile TrustManagerProvider trustmanagerProvider;
+
+    private ThreadPoolExecutor threadPool;
+    private HttpClient sharedHttpClient;
+    private HttpClientFacade sharedHttpClientFacade;
+
+    private int minThreadsShared;
+    private int maxThreadsShared;
+    private int keepAliveTimeoutShared;
+    private int minThreadsCustom;
+    private int maxThreadsCustom;
+    private int keepAliveTimeoutCustom;
+
+    @Activate
+    protected void activate(Map<String, Object> parameters) {
+        getConfigParameters(parameters);
+        threadPool = createThreadPool("common", minThreadsShared, maxThreadsShared, keepAliveTimeoutShared);
+
+        sharedHttpClient = createHttpClientInternal("common", null);
```

Comment:
Why not defer the creation until the first consumer actually requests it?

Reply:
I like the idea. We could also defer the creation of thread pool as well.

Reply:
I think this is a kind of over-engineering ;-) This is not an optional service, which might be used by someone. Especially for the thread pool it makes no sense, because the thread pool doesn't start threads before it becomes used.

Reply:
> This is not an optional service, which might be used by someone.

It is not? What part of ESH core does require it mandatorily?

Reply:
Isn't jupnp part of ESH core?

Reply:
No, JUPnP is a separate project which anyhow uses its separate thread pool.

Reply:
Ok, I do not insist on that.